### PR TITLE
Dataset preprocessing refactor to use event dataclasses and utilities.

### PR DESCRIPTION
**Pull request by harrykeightley**
## Preprocessing before this PR
Before this PR, the preprocessing work flow was as follows:
- Trigger the `process_dataset` function any time a dataset was created in firestore.
- This function would call indirectly call `process_file` on each file in the dataset (through publishing events to a pubsub topic)
- The processed transcription files here could contain multiple annotations for the same audio file
- All that happened to audio files was that they were resampled to 16k. 
- To check that the dataset had finished processing, at the end of `process_file`, we checked that the length of the files in the processed dataset folder in cloud storage was equal to the length of the files in the dataset (in firestore)

Down the line in the trainer service, the fact that a transcription file could have multiple utterances/annotations within it actually caused some annoyances- in that we had to do some additional preprocessing in the cloud run service that really should have been done in the cloud functions. As the cloud run service will have a beefy gpu, this would end up being more expensive for users (if we had any, or a product). 

Additionally, there was quite a severe bug where because elan annotation data was extracted before the audio files were resampled, and the annotation timstamps weren't updated in the preprocessing step, the annotation times would just be invalid unless the audio file was already at 16k before processing. 

## The goal
Generate a clipped audio file for each utterance in the preprocessing steps for cloud functions so that our dataset is all pretty by the time we get to the training service.

To do this, the process_file workflow needs to change to be aware of the corresponding audio file for each original transcription file.

### `functions/datasets/process_dataset.py`
- Use the refactored pubsub publish to topic utilities to drastically simplify the `process_dataset` function.
- Build a `Dataset` from the firestore event, which is also a lot simpler thanks to an unpacking utility.
- Generate a batch of `ProcessingJobs` from a dataset instance, which matches transcription and audio files before handing them to `process_file.py`. 

### `functions/datasets/process_file.py`
- Refactored what was a rabbit hole of functions (that I wrote) into a simpler, more functional workflow:
  - Extract annotations from transcription file
  - Resample the audio and annotations to be standardized at 16k 
  - Clean the annotation transcripts
  - Generate one processed transcription and audio file per annotation
  - Upload all these files
  - Tell the dataset in firestore that it's finished processing

### `utils/audio.py`
- Added utilities for cutting and getting sample rates from audio.

### Other
- Wrote lots of tests.

### CODE REVIEW

File: functions/models/dataset.py [hunk 1]
```diff
@@ -54,6 +59,32 @@ def to_dict(self) -> Dict[str, Any]:
...
+    audio_file_name: str
+    dataset_name: str
+    options: DatasetOptions
+
+    def to_dict(self) -> Dict[str, Any]:
+        result = dict(self.__dict__)
+        result["options"] = self.options.to_dict()
+        return result
+
+    @staticmethod
```
Comment (nicklambourne) [inline functions/models/dataset.py:77]: Could be `@classmethod`

**Review by nicklambourne — APPROVED** [review]
LGTM w/ some notes

File: functions/README.md [hunk 2]
```diff
@@ -22,3 +22,33 @@ with:
...
+If you get an error message on Mac Os that reads along the lines of:
+
+```
+OSError: cannot load library ... blah blah ... libsndfile.dylib' (no such file)
+```
+
+Here are the steps I took to fix this:
+
+```
+% brew install libsndfile
```
[the remark this answers is not in the document]
Comment (harrykeightley) [reply functions/README.md]: Fixed in latest revision.

File: functions/functions/datasets/process_file.py [hunk 3]
```diff
@@ -33,185 +33,189 @@ def process_dataset_file(event, context) -> None:
...
+) -> Tuple[Path, Path]:
+    """Generates a transcript and audio file pairing for this annotation.
 
-def clean_data(
-    annotations: List[Annotation], options: Dict[str, Any]
-) -> List[Annotation]:
-    """Takes a list of utterance information and cleans it based off the provided
-    dataset processing options.
+    If the annotation is timed (has a start and stop time), we return a path
+    to a new audio file, which is constrained to the given times.
```
[the remark this answers is not in the document]
Comment (harrykeightley) [reply functions/functions/datasets/process_file.py]: Fixed in latest revision.

File: functions/utils/audio.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+from pathlib import Path
+
+import soundfile as sf
```
[the remark this answers is not in the document]
Comment (harrykeightley) [reply functions/utils/audio.py:3]: Yeah I'm a big fan :) I don't think it has all the audio formats but more than enough.

File: functions/models/dataset.py (hunk 1 above)
Comment (harrykeightley) [reply functions/models/dataset.py:77]: Going to address in a future PR :)

**Pull request merged**